File: 704.py
class Solution(object):
    def search(self, nums: list[int], target: int) -> int:
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        left, right = 0, len(nums) - 1

        while left <= right :
            middle = left + (right - left) // 2

            if nums[middle] > target :
                right = middle - 1
            elif (nums[middle] < target):
                left = middle + 1
            else:
                return middle
        return -1

    def search2(self, nums, target):
        left, right = 0, len(nums)

        while left < right:
            middle = left + (right - left) // 2

            if nums[middle] > target:
                right = middle
            elif nums[middle] < target:
                left = middle + 1
            else:
                return middle
        return -1
        

if __name__ == '__main__':
    solution = Solution()
    print(solution.search([-1,0,3,5,9,12], 9))
    print(solution.search([-1,0,3,5,9,12], 2))
    print(solution.search2([-1,0,3,5,9,12], 9))
    print(solution.search2([-1,0,3,5,9,12], 2))

File: 0704.Binary Search/704.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;
        while (left <= right) {
            int middle = left + ((right - left) / 2);
            if (nums[middle] > target) {
                right = middle - 1;
            } else if (nums[middle] < target) {
                left = middle + 1;
            } else {
                return middle;
            }
        }
        return -1;
    }
    int search2(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;
        while (left < right) {
            int middle = left + ((right - left) / 2);
            if (nums[middle] > target) {
                right = middle;
            } else if (nums[middle] < target) {
                left = middle + 1;
            } else {
                return middle;
            }
        }
        return -1;
    }
};

int main ()
{
    Solution s;
    vector<int> v{-1, 0, 3, 5, 9, 12};
    int ans = s.search2(v, 9);
    cout << ans << endl;

    return 0;
}
